Fixes parameter inconsistency in IFieldWrapper (Fixes #4325)

thilokru opened the pull request:
As described in Issue #4325, the interface IFieldWrapper wasn't clear about what the method parameters ment. This caused issues with the MapWrapper, because it disagreed with ConfigManager about what should be in the parameter name.

This PR fixes the parameter naming inconsistency by renaming it to `key` (which also fixes the collision with the field `name`).
To prevent similar problems in the future, `IFieldWrapper` and `ITypeAdapter` were documented.
Also, `ConfigManager::sync` now has more detailed comments to allow other developers to easier understand the code (I had trouble reading my own code...)

Additionally, as indicated by @Choonster , Map entries weren't synced properly. This was caused by faulty return values for the category and the FieldWrapper ignoring the fact, that category names are all lowercase. This is fixed with 3a519a1.

Commit:
Added documentation, fixes inconsistencies and related bug.

The interfaces IFieldWrapper and ITypeAdapter were poorly documented, which was changed in this commit.
Especially IFieldWrapper was problematic, as the fully qualified property name was given as parameter "name" or "key" (-> inconsistency). Also, sometimes these parameters ment the suffix or the fully qualified name. This was fixed.
The related parts in ConfigManager were adjusted to these changes.

ConfigManager::sync was further documented, so that other programmers will have an easier time to debug and fix this code.

Commit:
Fixes formatting

Guidelines are brackets on new line.

thilokru commented:
@Choonster I'm pretty sure this fixes your issue, but please take a look to ensure it really does.

Choonster commented:
After some testing, I don't think this fixes #4325.

In TestMod3, I have a `Map<String, Double>` field in my config class that's populated with three default values at startup ([here](https://github.com/Choonster-Minecraft-Mods/TestMod3/blob/2ce586d2a90443f3c71d4c201cfeb7f5a4169888/src/main/java/choonster/testmod3/config/ModConfig.java#L23-L30)).

If I add a new property to the corresponding category in the config file and then launch the game, this new property is visible in the config GUI but not present in the `Map` (which I verified using [this command](https://github.com/Choonster-Minecraft-Mods/TestMod3/blob/2ce586d2a90443f3c71d4c201cfeb7f5a4169888/src/main/java/choonster/testmod3/command/CommandDebugConfig.java#L28-L32)).

thilokru commented:
Are you sure, that the way you use `@Mod.EventBusSubscriber` is correct? What you are describing sounds an awefull lot like `ConfigManager::sync` isn't been called when required. Maybe your event listener isn't working?

Choonster commented:
The event handler is definitely being called after changes have been made in the config GUI, but that's not really relevant as the new property is already present in the `Configuration` object when `ConfigManager.sync` is first called from `FMLModContainer#constructMod`.

thilokru commented:
I think I found out what caused this: The `MapWrapper` contains an implementation error; The base name is incorrect if the property has no parent category. Will fix this.

Commit:
Fixes map syncing bug caused by invalid case

The categories are always lower case, this changes the MapWrapper to behave accordingly. It also fixes a bug caused by invalid category names when the maps base category is empty.
ConfigTest gets a new section testing map key syncing.

thilokru commented:
This should fix the original bug and the problems reported above.

LexManos commented on src/main/java/net/minecraftforge/common/config/ConfigManager.java:
```diff
@@ -218,39 +218,47 @@ private static void sync(Configuration cfg, Class<?> cls, String modid, String c
         {
-            if (!Modifier.isPublic(f.getModifiers()))
+            if (!Modifier.isPublic(f.getModifiers())) //Allow modders to have private field for internal data
                 continue;
```
Remove all your useless comment spam.

thilokru replied:
I'll gladly do, but when I tried to understand this code (which I wrote almost a year ago), I could barely understand it anymore. As I suspect that there will be more bugs and/or missing features, other people might want to contribute. I think it is better to comment the code so that strangers can understand.

However, if you insist, I will remove it all.

LexManos replied:
The only confusion is that you didnt clearly specify what the key should be.
But things like 'getType() Returns the type!' is just redundant.
Your code should be self-documenting for the most part sprinkeled with a few "I couldn't specify the format of the key in the variable name so here is the correct format" style comments.

thilokru replied:
OK, I will try to change the comments to fit your criteria.

Commit:
Cleaned up comments

Removed too obvious comments (i.e. in ITypeAdapter getType() doesn't need a comment) and reduced the comments in ConfigManager::sync to a bare minimum, only explaining behaviour not obvious from the code, like what kinds of wrappers exist, under which circumstances members must be static, or how it is decided in which direction data is moved (disk<->Configuration Object <-> Config Class)

thilokru commented:
Reson why this PR isn't stale: I need to fix the merge conflict, but currently, I don't have access to my computer to do that. Within a week I should be able to fix this.

Commit:
Merge remote-tracking branch 'remotes/upstream/1.12.x' into 1.12.x

thilokru commented:
Merge conflicts have been resolved.
